File: src/logic/segments.js
const User = require("../models/userData/user");
const { getQuery } = require("./querys");
const AWS = require('aws-sdk');
const csv = require('csvtojson');


async function getUserSegment(conditions, limitUsers, testUser) {

    if (testUser) {
        let users = await User.find({ "msisdn": testUser });
        return users;
    }

    let querys = [];
    conditions.forEach(conditionGroup => {
        let tempQuery = conditionGroup.map(condition => {
            return getQuery(condition);
        }).join(",");
        querys.push(`{"$and": [${tempQuery}]}`);
    });
    let query = `{"$or": [${querys.join(",")}]}`;
    let objectQuery = JSON.parse(query);

    if (limitUsers) {
        let users = await User.find(objectQuery).limit(limitUsers);
        return users;
    }

    let users = await User.find(objectQuery);
    return users;

}

async function getUserFromBucket(path) {
    try {
        if (path == null) {
            return []
        }

        var s3 = new AWS.S3();
        let data = path.split("//");
        let bucket = data[1].split("/")[0];
        let key = path.split(bucket + "/")[1];

        var getParams = {
            Bucket: bucket, // your bucket name,
            Key: key // path to the object you're looking for
        }
        let stream = s3.getObject(getParams).createReadStream()
        // .on('error', (err) => reject(err))
        // .on('end', () => resolve());;

        // let stream = await new Promise((resolve, reject) => {
        //     const params = {
        //         Bucket: bucket, // your bucket name,
        //         Key: key // path to the object you're looking for
        //     }
        //     s3.getObject(params)
        //         .createReadStream()
        //         .on('error', (err) => reject(err))
        //         .on('end', () => resolve());
        // });

        const json = await csv().fromStream(stream);
        // console.log(json);
        return json;
    } catch (err) {
        console.error(err);
        return []
    }
}


module.exports = { getUserSegment, getUserFromBucket }

File: py_logic/python_lambdas/segmentation-engine-core-python-evaluation/segments.py
import boto3
import csv
import json
def get_user_from_bucket(path):
    try:
        print("path", path)
        if path is None:
            return []

        s3_client = boto3.client('s3')

        data = path.split("//")
        bucket = data[1].split("/")[0]
        key = path.split(bucket + "/")[1]
        
        print("bucket", bucket)
        print("key", key)

        # Get the file inside the S3 Bucket
        s3_response = s3_client.get_object(
            Bucket= bucket,
            Key= key
        )
        print("s3_response", s3_response)
        # Get the Body object in the S3 get_object() response
        s3_object_body = s3_response.get('Body')
        print("s3_object_body", s3_object_body)
        read = csv.DictReader(s3_object_body.read().decode('utf-8').splitlines())
        users = list(read)
        return users

    except Exception as e:
        print("error datos", e)
        return []